From 1ad0ab33bc2a094aae890b5790aa89fa7396fc65 Mon Sep 17 00:00:00 2001
From: J. Bruce Fields <bfields@redhat.com>
Date: Tue, 16 Sep 2014 20:18:08 -0400
Subject: [fs] nfsd: Using exp_get for export getting

Message-id: <1410898849-22126-5-git-send-email-bfields@redhat.com>
Patchwork-id: 95845
O-Subject: [Patch RHEL7 004/165] NFSD: Using exp_get for export getting
Bugzilla: 1044798 1123721 1141932
RH-Acked-by: Benjamin Coddington <bcodding@redhat.com>
RH-Acked-by: Steve Dickson <SteveD@redhat.com>

From: Kinglong Mee <kinglongmee@gmail.com>

Don't using cache_get besides export.h, using exp_get for export.

Signed-off-by: Kinglong Mee <kinglongmee@gmail.com>
Signed-off-by: J. Bruce Fields <bfields@redhat.com>
Upstream: bf18f163e89c52e09c96534db45c4274273a0b34
Bugzilla: 1044798
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/fs/nfsd/export.c b/fs/nfsd/export.c
index ef2d9d6..72ffd7c 100644
--- a/fs/nfsd/export.c
+++ b/fs/nfsd/export.c
@@ -1253,7 +1253,7 @@ static int e_show(struct seq_file *m, void *p)
   return 0;
  }
 
- cache_get(&exp->h);
+ exp_get(exp);
  if (cache_check(cd, &exp->h, NULL))
   return 0;
  exp_put(exp);
diff --git a/fs/nfsd/export.h b/fs/nfsd/export.h
index cfeea85..04dc8c1 100644
--- a/fs/nfsd/export.h
+++ b/fs/nfsd/export.h
@@ -101,9 +101,10 @@ static inline void exp_put(struct svc_export *exp)
  cache_put(&exp->h, exp->cd);
 }
 
-static inline void exp_get(struct svc_export *exp)
+static inline struct svc_export *exp_get(struct svc_export *exp)
 {
  cache_get(&exp->h);
+ return exp;
 }
 struct svc_export * rqst_exp_find(struct svc_rqst *, int, u32 *);
 
diff --git a/fs/nfsd/nfs4proc.c b/fs/nfsd/nfs4proc.c
index e8dff17..3ece7cd 100644
--- a/fs/nfsd/nfs4proc.c
+++ b/fs/nfsd/nfs4proc.c
@@ -176,7 +176,7 @@ fh_dup2(struct svc_fh *dst, struct svc_fh *src)
  fh_put(dst);
  dget(src->fh_dentry);
  if (src->fh_export)
-  cache_get(&src->fh_export->h);
+  exp_get(src->fh_export);
  *dst = *src;
 }
 
@@ -917,8 +917,8 @@ nfsd4_secinfo_no_name(struct svc_rqst *rqstp, struct nfsd4_compound_state *cstat
  default:
   return nfserr_inval;
  }
- exp_get(cstate->current_fh.fh_export);
- sin->sin_exp = cstate->current_fh.fh_export;
+
+ sin->sin_exp = exp_get(cstate->current_fh.fh_export);
  fh_put(&cstate->current_fh);
  return nfs_ok;
 }
diff --git a/fs/nfsd/nfsfh.c b/fs/nfsd/nfsfh.c
index ec83934..6f5cc76 100644
--- a/fs/nfsd/nfsfh.c
+++ b/fs/nfsd/nfsfh.c
@@ -539,8 +539,7 @@ fh_compose(struct svc_fh *fhp, struct svc_export *exp, struct dentry *dentry,
          dentry);
 
  fhp->fh_dentry = dget(dentry); /* our internal copy */
- fhp->fh_export = exp;
- cache_get(&exp->h);
+ fhp->fh_export = exp_get(exp);
 
  if (fhp->fh_handle.fh_version == 0xca) {
   /* old style filehandle please */
diff --git a/fs/nfsd/vfs.c b/fs/nfsd/vfs.c
index 762f716..4c21643 100644
--- a/fs/nfsd/vfs.c
+++ b/fs/nfsd/vfs.c
@@ -189,8 +189,7 @@ nfsd_lookup_dentry(struct svc_rqst *rqstp, struct svc_fh *fhp,
  dprintk("nfsd: nfsd_lookup(fh %s, %.*s)\n", SVCFH_fmt(fhp), len,name);
 
  dparent = fhp->fh_dentry;
- exp  = fhp->fh_export;
- exp_get(exp);
+ exp = exp_get(fhp->fh_export);
 
  /* Lookup the name, but don't follow links */
  if (isdotent(name, len)) {
-- 
1.7.1